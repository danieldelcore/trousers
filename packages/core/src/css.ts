import hash from '@trousers/hash';

import { Definition, CSSObject } from './types';
import themify from './themify';

export interface CollectorReturn {
    _get: () => Definition[];
    modifier: (id: string, styles: CSSObject) => CollectorReturn;
    global: (styles: CSSObject) => CollectorReturn;
    theme: (theme: Record<string, any>) => CollectorReturn;
}

function css(styles: CSSObject): CollectorReturn;
function css(
    idOrStyle: string | CSSObject,
    styles?: CSSObject,
): CollectorReturn;
function css(
    idOrStyle: string | CSSObject,
    styles?: CSSObject,
): CollectorReturn {
    let id = typeof idOrStyle === 'string' ? idOrStyle : '';
    let styleObject = (typeof idOrStyle === 'string'
        ? styles
        : idOrStyle) as CSSObject;

    const elementId = `${id}-${hash(JSON.stringify(styleObject))}`;
    const styleMap: Definition[] = [
        {
            id,
            type: 'element',
            className: `.${elementId}`,
            styles: styleObject,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                id: modifierId,
                type: 'modifier',
                className: `.${elementId}--${modifierId}-${hash(
                    JSON.stringify(modifierStyles),
                )}`,
                styles: modifierStyles,
            });

            return self;
        },
        global: globalStyles => {
            styleMap.push({
                id: `global-${id}-${hash(JSON.stringify(globalStyles))}`,
                type: 'global',
                className: '',
                styles: globalStyles,
            });

            return self;
        },
        theme: theme => {
            const themeHash = hash(JSON.stringify(theme));
            styleMap.push({
                id: `theme-${id}-${themeHash}`,
                type: 'theme',
                className: `.theme-${id}-${themeHash}`,
                styles: themify(theme),
            });

            return self;
        },
    };

    return self;
}

export type Collector = typeof css;

export default css;
