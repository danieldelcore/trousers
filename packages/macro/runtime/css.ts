interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: string;
}

interface CollectorReturn {
    modifier: (id: string, styles: string) => CollectorReturn;
    global: (id: string, styles: string) => CollectorReturn;
    theme: (id: string, styles: string) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: string, elementStyles: string) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
