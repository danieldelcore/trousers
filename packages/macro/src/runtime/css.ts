interface Definition {
    type: 'element' | 'modifier' | 'global' | 'theme';
    id: string;
    styles: Record<string, string>;
}

export interface CollectorReturn {
    modifier: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    global: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    theme: (
        id: Definition['id'],
        styles: Definition['styles'],
    ) => CollectorReturn;
    _get: () => Definition[];
}

function css(elementId: Definition['id'], elementStyles: Definition['styles']) {
    const styleMap: Definition[] = [
        {
            type: 'element',
            id: elementId,
            styles: elementStyles,
        },
    ];

    const self: CollectorReturn = {
        _get: () => styleMap,
        modifier: (modifierId, modifierStyles) => {
            styleMap.push({
                type: 'modifier',
                id: modifierId,
                styles: modifierStyles,
            });

            return self;
        },
        global: (globalId, globalStyles) => {
            styleMap.push({
                type: 'global',
                id: globalId,
                styles: globalStyles,
            });

            return self;
        },
        theme: (themeId, themeStyles) => {
            styleMap.push({
                type: 'theme',
                id: themeId,
                styles: themeStyles,
            });

            return self;
        },
    };

    return self;
}

export default css;
