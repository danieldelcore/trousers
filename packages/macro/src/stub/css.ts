/* eslint-disable @typescript-eslint/no-unused-vars */
import {
    CSSObject,
    Collector as CoreCollector,
    CollectorReturn as CoreCollectorReturn,
} from '@trousers/core';

export type CollectorReturn = CoreCollectorReturn;
export type Collector = CoreCollector;

function css(_styles: CSSObject): CollectorReturn;
function css(
    _idOrStyle: string | CSSObject,
    _styles?: CSSObject,
): CollectorReturn;
function css(
    _idOrStyle: string | CSSObject,
    _styles?: CSSObject,
): CollectorReturn {
    const self: CollectorReturn = {
        modifier: (_modifierId, _modifierStyles) => self,
        global: _globalStyles => self,
        theme: _theme => self,
        _get: () => [],
    };

    return self;
}

export default css;
