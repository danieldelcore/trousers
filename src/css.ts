import { StyleCollector } from './';
import { Expression, StyleDefinition } from './types';

export interface SingleStyleCollector<Theme = {}> {
    get: () => StyleDefinition<Theme>[];
}

export default function css<Theme = {}>(
    styles: TemplateStringsArray,
    ...expressions: Expression<Theme>[]
): SingleStyleCollector<Theme> {
    const styleCollector = new StyleCollector<Theme>('css');

    styleCollector.element(styles, ...expressions);

    return {
        get: () => styleCollector.get(),
    };
}
