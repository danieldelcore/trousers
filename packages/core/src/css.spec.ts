import css from './css';

describe('css', () => {
    it('collects id and element styles', () => {
        expect(css('Element', { color: 'red' })._get()).toMatchSnapshot();
    });

    it('collects element styles without an id supplied', () => {
        expect(css({ color: 'red' })._get()).toMatchSnapshot();
    });

    it('collects single modifier style', () => {
        expect(
            css('Element', { color: 'red' })
                .modifier('modifier', { color: 'blue' })
                ._get(),
        ).toMatchSnapshot();
    });

    it('collects multiple modifier styles', () => {
        expect(
            css('Element', { color: 'red' })
                .modifier('primary', { color: 'blue' })
                .modifier('secondary', { color: 'grey' })
                ._get(),
        ).toMatchSnapshot();
    });

    it('collects single global style', () => {
        expect(
            css('Element', {})
                .global({
                    ':root': {
                        color: 'green',
                    },
                })
                ._get(),
        ).toMatchSnapshot();
    });

    it('collects multiple global styles', () => {
        expect(
            css('Element', { color: 'red' })
                .global({
                    ':root': {
                        color: 'green',
                    },
                })
                .global({
                    '*': {
                        boxSizing: 'border-box',
                    },
                })
                ._get(),
        ).toMatchSnapshot();
    });

    it('collects single flat theme object', () => {
        expect(
            css('Element', { color: 'red' })
                .theme({
                    color: 'red',
                    background: 'blue',
                })
                ._get(),
        ).toMatchSnapshot();
    });

    it('collects single camel case flat theme object', () => {
        expect(
            css('Element', { color: 'red' })
                .theme({
                    primaryColor: 'red',
                    primaryBackground: 'blue',
                })
                ._get(),
        ).toMatchSnapshot();
    });

    it('collects single deep theme object', () => {
        expect(
            css('Element', { color: 'red' })
                .theme({
                    color: {
                        foreground: 'red',
                        background: 'blue',
                    },
                })
                ._get(),
        ).toMatchSnapshot();
    });

    it('collects multiple theme objects', () => {
        expect(
            css('Element', { color: 'red' })
                .theme({
                    typography: {
                        size: '18px',
                        font: 'ariel',
                    },
                })
                .theme({
                    color: {
                        foreground: 'red',
                        background: 'blue',
                    },
                })
                ._get(),
        ).toMatchSnapshot();
    });
});
