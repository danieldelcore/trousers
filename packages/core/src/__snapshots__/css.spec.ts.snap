// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css collects element styles without an id supplied 1`] = `
Array [
  Object {
    "className": ".2313942302",
    "id": "",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
]
`;

exports[`css collects id and element styles 1`] = `
Array [
  Object {
    "className": ".Element-2313942302",
    "id": "Element",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
]
`;

exports[`css collects multiple global styles 1`] = `
Array [
  Object {
    "className": ".Element-2313942302",
    "id": "Element",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
  Object {
    "className": "",
    "id": "global-Element-3713844026",
    "styles": Object {
      ":root": Object {
        "color": "green",
      },
    },
    "type": "global",
  },
  Object {
    "className": "",
    "id": "global-Element-2840557653",
    "styles": Object {
      "*": Object {
        "boxSizing": "border-box",
      },
    },
    "type": "global",
  },
]
`;

exports[`css collects multiple modifier styles 1`] = `
Array [
  Object {
    "className": ".Element-2313942302",
    "id": "Element",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
  Object {
    "className": ".Element-2313942302--primary-2561700995",
    "id": "primary",
    "styles": Object {
      "color": "blue",
    },
    "type": "modifier",
  },
  Object {
    "className": ".Element-2313942302--secondary-2709930440",
    "id": "secondary",
    "styles": Object {
      "color": "grey",
    },
    "type": "modifier",
  },
]
`;

exports[`css collects multiple theme objects 1`] = `
Array [
  Object {
    "className": ".Element-2313942302",
    "id": "Element",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
  Object {
    "className": ".theme-Element-4166118473",
    "id": "theme-Element-4166118473",
    "styles": Object {
      "--typography-font": "ariel",
      "--typography-size": "18px",
    },
    "type": "theme",
  },
  Object {
    "className": ".theme-Element-1320425247",
    "id": "theme-Element-1320425247",
    "styles": Object {
      "--color-background": "blue",
      "--color-foreground": "red",
    },
    "type": "theme",
  },
]
`;

exports[`css collects single camel case flat theme object 1`] = `
Array [
  Object {
    "className": ".Element-2313942302",
    "id": "Element",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
  Object {
    "className": ".theme-Element-1395378648",
    "id": "theme-Element-1395378648",
    "styles": Object {
      "--primary-background": "blue",
      "--primary-color": "red",
    },
    "type": "theme",
  },
]
`;

exports[`css collects single deep theme object 1`] = `
Array [
  Object {
    "className": ".Element-2313942302",
    "id": "Element",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
  Object {
    "className": ".theme-Element-1320425247",
    "id": "theme-Element-1320425247",
    "styles": Object {
      "--color-background": "blue",
      "--color-foreground": "red",
    },
    "type": "theme",
  },
]
`;

exports[`css collects single flat theme object 1`] = `
Array [
  Object {
    "className": ".Element-2313942302",
    "id": "Element",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
  Object {
    "className": ".theme-Element-3131109692",
    "id": "theme-Element-3131109692",
    "styles": Object {
      "--background": "blue",
      "--color": "red",
    },
    "type": "theme",
  },
]
`;

exports[`css collects single global style 1`] = `
Array [
  Object {
    "className": ".Element-3938",
    "id": "Element",
    "styles": Object {},
    "type": "element",
  },
  Object {
    "className": "",
    "id": "global-Element-3713844026",
    "styles": Object {
      ":root": Object {
        "color": "green",
      },
    },
    "type": "global",
  },
]
`;

exports[`css collects single modifier style 1`] = `
Array [
  Object {
    "className": ".Element-2313942302",
    "id": "Element",
    "styles": Object {
      "color": "red",
    },
    "type": "element",
  },
  Object {
    "className": ".Element-2313942302--modifier-2561700995",
    "id": "modifier",
    "styles": Object {
      "color": "blue",
    },
    "type": "modifier",
  },
]
`;
